source/framework/vx_array.c

```diff
@@ -22,6 +22,129 @@ static void ownInitArrayObject(
     vx_array arr, vx_enum item_type, vx_size capacity, vx_bool is_virtual);
 static vx_size ownGetArrayItemSize(vx_context context, vx_enum item_type);
 static vx_bool ownIsValidArrayItemType(vx_context context, vx_enum item_type);
+static vx_bool ownIsValidInputAndOutputArrays(vx_array input, vx_array output);
+static vx_status isArrayCopyable(vx_array input, vx_array output);
+static vx_status VX_CALLBACK arrayKernelCallback(vx_enum kernel_enum, vx_bool validate_only, const vx_reference params[], vx_uint32 num_params);
+
+/*! \brief checks that the input and output references are not the same, and that both are valid arrays
+*/
+static vx_bool ownIsValidInputAndOutputArrays(vx_array input, vx_array output)
+{
+    vx_bool res = (vx_bool)vx_false_e;
+    if ((input != output) &&
+        (ownIsValidSpecificReference(&input->base, (vx_enum)VX_TYPE_ARRAY) == (vx_bool)vx_true_e) &&
+        (input->base.obj_desc != NULL) &&
+        (ownIsValidSpecificReference(&output->base, (vx_enum)VX_TYPE_ARRAY) == (vx_bool)vx_true_e) &&
+        (output->base.obj_desc != NULL)
+        )
+    {
+        res = (vx_bool)vx_true_e;
+    }
+    return res;
+}
+
+/*! \brief This function is called to find out if it is OK to copy the input to the output.
+ * If the output is virtual, a zero capacity or invalid item type is OK
+ * Will also copy metadata from input to output.
+ * \returns VX_SUCCESS if it is, otherwise another error code.
+ *
+ */
+static vx_status isArrayCopyable(vx_array input, vx_array output)
+{
+    vx_status status = (vx_status)VX_SUCCESS;
+    tivx_obj_desc_array_t *ip_obj_desc = (tivx_obj_desc_array_t *)input->base.obj_desc;
+    tivx_obj_desc_array_t *op_obj_desc = (tivx_obj_desc_array_t *)output->base.obj_desc;
+    if ((vx_bool)vx_false_e == ownIsValidInputAndOutputArrays(input, output))
+    {
+        status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+    }
+    else if ((vx_bool)vx_true_e == output->base.is_virtual)
+    {
+        /* Either output type must be invalid or types must match*/
+        if (((vx_enum)VX_TYPE_INVALID != op_obj_desc->item_type) &&
+            (ip_obj_desc->item_type != op_obj_desc->item_type))
+        {
+            status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+        }
+        /* Either output capacity must be zero or at least as large as input capacity */
+        if ((0U != op_obj_desc->capacity) &&
+             (ip_obj_desc->capacity != op_obj_desc->capacity))
+        {
+            status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+        }
+    }
+    else if (op_obj_desc->item_type != ip_obj_desc->item_type)
+    {
+        /* Types must match */
+        status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+    }
+    else if (op_obj_desc->capacity < ip_obj_desc->capacity)
+    {
+        /* Output must have sufficient capacity to hold input */
+        status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+    }
+    else
+    {
+        op_obj_desc->item_type = ip_obj_desc->item_type;
```

Comment:
Is the "num_items" attribute intentionally excluded from the copy to the op_obj_desc?  We have been debating internally on this one and wanted to see if you had thoughts on this

Reply:
the num_items is copied in the generic ownCopyReferenceGeneric function. there is a special case that handles the vx_array type. To be honest, I would refer to include the full copy inside the ownCopyReferenceGeneric function, because the isArrayCopyable should not modify the data (I wanted to stick to the original implementation)
I cannot use here the generic functions as the conditions are a bit different to the tivxIsReferenceMetaFormatEqual function:
- capacity of the output has to be >= than the input
- in case of virtual object the  input and output have to have the same size and different than 0

Reply:
As discussed in call, comments will be added to the code to clarify the need for this